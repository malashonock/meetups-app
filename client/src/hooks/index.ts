export * from './useMeetupQuery';
export * from './useNewsArticleQuery';
export * from './useStaticFileQuery';
export * from './useAuthStore';
