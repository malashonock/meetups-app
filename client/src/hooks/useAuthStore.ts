import { useContext } from 'react';

import { AuthStore } from 'stores';
import { RootContext } from 'components';

export const useAuthStore = (): AuthStore => {
  return useContext(RootContext).authStore;
};
