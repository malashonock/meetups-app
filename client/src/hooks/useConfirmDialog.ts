import { useContext } from 'react';

import { ConfirmDialogContext, ConfirmFn } from 'components';
import { Optional } from 'types';

export const useConfirmDialog = (): ConfirmFn => {
  return useContext(ConfirmDialogContext);
};
