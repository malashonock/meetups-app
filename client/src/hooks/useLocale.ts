import { useContext } from 'react';

import { UiStore, Locale } from 'stores';
import { RootContext } from 'components';
import { Optional } from 'types';

export type UseLocaleResult = [
  locale: Optional<Locale>,
  setLocale: (locale: Locale) => void,
];

export const useLocale = (): UseLocaleResult => {
  const uiStore: Optional<UiStore> = useContext(RootContext)?.uiStore;

  const locale = uiStore?.locale;

  const setLocale = (locale: Locale): void => {
    if (uiStore) {
      uiStore.locale = locale;
    }
  };

  return [locale, setLocale];
};
