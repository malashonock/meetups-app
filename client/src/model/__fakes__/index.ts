export * from './user.fake';
export * from './meetup.fake';
export * from './file.fake';
