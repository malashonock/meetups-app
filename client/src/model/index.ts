export * from './meetup.model';
export * from './news.model';
export * from './user.model';
export * from './login.model';
