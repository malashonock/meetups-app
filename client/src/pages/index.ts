export * from './LoginPage/LoginPage';
export * from './MeetupsPage/MeetupsPage';
export * from './ViewMeetupPage/ViewMeetupPage';
export * from './CreateMeetupPage/CreateMeetupPage';
export * from './EditMeetupPage/EditMeetupPage';
export * from './NewsPage/NewsPage';
export * from './ViewNewsPage/ViewNewsPage';
export * from './CreateNewsPage/CreateNewsPage';
export * from './EditNewsPage/EditNewsPage';
export * from './NotFoundPage/NotFoundPage';
