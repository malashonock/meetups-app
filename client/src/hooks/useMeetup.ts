import { useEffect } from 'react';

import { Meetup } from 'stores';
import { Maybe, Optional } from 'types';
import { useMeetupStore } from 'hooks';

export const useMeetup = (id: Maybe<string>): Optional<Meetup> => {
  const meetupStore = useMeetupStore();

  const meetup = id ? meetupStore.findMeetup(id) : undefined;

  // Hydrate meetup on first load
  useEffect((): void => {
    (async (): Promise<void> => {
      if (meetup && !meetup?.isInitialized) {
        await meetup?.init();
      }
    })();
  }, [meetup, meetup?.isInitialized]);

  return meetup;
};
