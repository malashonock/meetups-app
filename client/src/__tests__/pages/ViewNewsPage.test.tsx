import { PropsWithChildren } from 'react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ViewNewsPage } from 'pages';
import { mockNewsArticle, mockUser } from 'model/__fakes__';
import { useAuthStore, useNewsArticle } from 'hooks';

// Mock useAuthStore & useNewsArticle hook
jest.mock('hooks', () => {
  return {
    ...jest.requireActual('hooks'),
    useAuthStore: jest.fn(),
    useNewsArticle: jest.fn(),
  };
});
const mockUseAuthStore = useAuthStore as jest.MockedFunction<
  typeof useAuthStore
>;
const mockUseNewsArticle = useNewsArticle as jest.MockedFunction<
  typeof useNewsArticle
>;

beforeEach(() => {
  mockUseAuthStore.mockReturnValue({
    loggedUser: mockUser,
  });
  mockUseNewsArticle.mockReturnValue({
    newsArticle: mockNewsArticle,
    isLoading: false,
    isError: false,
    errors: [],
  });
});

afterEach(() => {
  jest.resetAllMocks();
});

const MockRouter = ({ children }: PropsWithChildren): JSX.Element => (
  <MemoryRouter initialEntries={['/news', '/news/aaa']}>
    <Routes>
      <Route path="/news">
        <Route index element={<h1>News page</h1>} />
        <Route path=":id">
          <Route index element={children} />
          <Route path="edit" element={<h1>Edit news article</h1>} />
        </Route>
      </Route>
    </Routes>
  </MemoryRouter>
);

describe('ViewNewsPage', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<ViewNewsPage />, { wrapper: MockRouter });
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Back button', () => {
    it('on click, should navigate to previous page', async () => {
      render(<ViewNewsPage />, { wrapper: MockRouter });
      userEvent.click(screen.getByText('formButtons.back'));
      expect(screen.getByText('News page')).toBeInTheDocument();
    });
  });

  describe('Edit News button', () => {
    describe('given no user is logged in', () => {
      beforeEach(() => {
        mockUseAuthStore.mockReturnValue({
          loggedUser: null,
        });
      });

      it('should not be rendered', () => {
        render(<ViewNewsPage />, { wrapper: MockRouter });
        expect(screen.queryByText('formButtons.edit')).toBeNull();
      });
    });

    describe('given a user is logged in', () => {
      it('on click, should redirect to Edit News page', async () => {
        render(<ViewNewsPage />, { wrapper: MockRouter });
        userEvent.click(screen.getByText('formButtons.edit'));
        expect(screen.getByText('Edit news article')).toBeInTheDocument();
      });
    });
  });

  it('should render a Loading spinner if news article is undefined', () => {
    mockUseNewsArticle.mockReturnValue({});
    render(<ViewNewsPage />, { wrapper: MockRouter });
    expect(screen.getByText('loadingText.newsArticle')).toBeInTheDocument();
  });

  it('should render a Loading spinner while news article is loading', () => {
    mockUseNewsArticle.mockReturnValue({
      newsArticle: mockNewsArticle,
      isLoading: true,
    });
    render(<ViewNewsPage />, { wrapper: MockRouter });
    expect(screen.getByText('loadingText.newsArticle')).toBeInTheDocument();
  });

  it('should render Not Found page if an error occurred while loading the news article', () => {
    mockUseNewsArticle.mockReturnValue({
      newsArticle: mockNewsArticle,
      isError: true,
    });
    render(<ViewNewsPage />, { wrapper: MockRouter });
    expect(screen.getByText('notFoundPage.title')).toBeInTheDocument();
  });
});
