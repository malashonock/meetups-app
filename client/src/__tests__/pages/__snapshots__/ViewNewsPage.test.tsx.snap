// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewNewsPage should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="container"
    data-testid="news-page"
  >
    <h1
      class="typography h1 heading"
    >
      viewNewsPage.title
    </h1>
    <div
      class="contentWrapper"
    >
      <figure
        class="section imageWrapper"
        data-testid="image"
      >
        <img
          alt="viewNewsPage.imgAlt"
          class="image"
          src="http://localhost/test.jpg"
        />
      </figure>
      <div
        class="textSection main"
      >
        <h2
          class="typography h2 title"
          data-testid="title"
        >
          Test news title
        </h2>
        <p
          class="typography p text"
          data-testid="text"
        >
          Test news text
        </p>
      </div>
      <div
        class="textSection actions"
      >
        <button
          class="actionButton backBtn button default"
          data-testid="btn-back"
          id="btn-back"
        >
          formButtons.back
        </button>
        <div
          class="actionGroup"
        >
          <button
            class="actionButton editBtn button secondary"
            data-testid="btn-edit"
            id="btn-edit"
          >
            formButtons.edit
          </button>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
